src/SlidingWindow.cpp

```diff
@@ -343,43 +469,83 @@ class SlidingWindowOnFunctionAndLoop : public IRMutator {
     SlidingWindowOnFunctionAndLoop(Function f, string v, Expr v_min)
         : func(std::move(f)), loop_var(std::move(v)), loop_min(std::move(v_min)) {
     }
-};
-
-// Perform sliding window optimization for a particular function
-class SlidingWindowOnFunction : public IRMutator {
-    Function func;
 
-    using IRMutator::visit;
-
-    Stmt visit(const For *op) override {
-        debug(3) << " Doing sliding window analysis over loop: " << op->name << "\n";
+    Expr new_loop_min;
+};
 
-        Stmt new_body = op->body;
+// In Stmt s, does the production of b depend on a?
+// We can't use produce/consume nodes to determine this, because they're "loose".
+// For example, we get this:
+//
+//  produce a {
+//   a(...) = ...
+//  }
+//  consume a {
+//   produce b {
+//    b(...) = ... // not depending on a
+//   }
+//   consume b {
+//    c(...) = a(...) + b(...)
+//   }
+//  }
+//
+// When we'd rather see this:
+//
+//  produce a {
+//   a(...) = ...
+//  }
+//  produce b {
+//   b(...) = ... // not depending on a
+//  }
+//  consume a {
+//   consume b {
+//    c(...) = a(...) + b(...)
+//   }
+//  }
+//
+// TODO: We might also need to figure out transitive dependencies...? If so, it
+// would be best to just fix the produce/consume relationships as above. We would
+// just be able to look for produce b inside produce a.
+class DependsOn : public IRVisitor {
+    using IRVisitor::visit;
 
-        new_body = mutate(new_body);
+    const Function &a;
+    const Function &b;
+    bool finding_a = false;
 
-        if (op->for_type == ForType::Serial ||
-            op->for_type == ForType::Unrolled) {
-            new_body = SlidingWindowOnFunctionAndLoop(func, op->name, op->min).mutate(new_body);
-        }
+    void visit(const ProducerConsumer *op) override {
+        ScopedValue<bool> old_finding_a(finding_a, op->is_producer && op->name == b.name());
+        return IRVisitor::visit(op);
+    }
 
-        if (new_body.same_as(op->body)) {
-            return op;
+    void visit(const Call *op) override {
+        if (finding_a && op->name == a.name()) {
+            yes = true;
         } else {
-            return For::make(op->name, op->min, op->extent, op->for_type, op->device_api, new_body);
+            IRVisitor::visit(op);
         }
     }
 
 public:
-    SlidingWindowOnFunction(Function f)
-        : func(std::move(f)) {
+    bool yes = false;
+
+    DependsOn(const Function &a, const Function &b)
+        : a(a), b(b) {
     }
 };
 
+bool depends_on(const Function &a, const Function &b, const Stmt &s) {
+    DependsOn check(a, b);
+    s.accept(&check);
+    return check.yes;
+}
+
 // Perform sliding window optimization for all functions
 class SlidingWindow : public IRMutator {
     const map<string, Function> &env;
 
+    list<Function> sliding;
```

Comment:
A comment here on what this holds would be nice. It is the realize nodes currently in scope for things we intend to slide, from innermost to outermost?